buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java

```diff
@@ -0,0 +1,189 @@
+import org.elasticsearch.gradle.testclusters.TestDistribution;
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.plugins.JavaBasePlugin;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+
+import java.io.File;
+import java.nio.file.Path;
+
+import static org.elasticsearch.gradle.test.rest.RestTestUtil.createTestCluster;
+import static org.elasticsearch.gradle.test.rest.RestTestUtil.setupDependencies;
+
+/**
+ * Apply this plugin to run the YAML based REST tests from a prior major version against this version's cluster.
+ */
+public class YamlRestCompatTestPlugin implements Plugin<Project> {
+
+    public static final String SOURCE_SET_NAME = "yamlRestCompatTest";
+    private static final Path RELATIVE_API_PATH = Path.of("rest-api-spec/api");
+    private static final Path RELATIVE_TEST_PATH = Path.of("rest-api-spec/test");
+
+    @Override
+    public void apply(Project project) {
+
+        project.getPluginManager().apply(ElasticsearchJavaPlugin.class);
+        project.getPluginManager().apply(TestClustersPlugin.class);
+        project.getPluginManager().apply(RestTestBasePlugin.class);
+        project.getPluginManager().apply(RestResourcesPlugin.class);
+        project.getPluginManager().apply(YamlRestTestPlugin.class);
+
+        RestResourcesExtension extension = project.getExtensions().getByType(RestResourcesExtension.class);
+
+        // create source set
+        SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+        SourceSet yamlCompatTestSourceSet = sourceSets.create(SOURCE_SET_NAME);
+        SourceSet yamlTestSourceSet = sourceSets.getByName(YamlRestTestPlugin.SOURCE_SET_NAME);
+        GradleUtils.extendSourceSet(project, YamlRestTestPlugin.SOURCE_SET_NAME, SOURCE_SET_NAME);
+
+        // create the test cluster container, and always use the default distribution
+        ElasticsearchCluster testCluster = createTestCluster(project, yamlCompatTestSourceSet);
+        testCluster.setTestDistribution(TestDistribution.DEFAULT);
+
+        // TODO: this is pretty fragile but the existing logic to checkout prior version branches isn't general purpose
+        // eventually we will want to test against multiple minor versions and will need a better way to checkout an arbitrary prior branch
+        // however, for now we will just get a reference to the checkout directory for ":distribution:bwc:minor:checkoutBwcBranch"
+        int priorMajorVersion = VersionProperties.getElasticsearchVersion().getMajor() - 1;
+        final Path checkoutDir = project.findProject(":distribution:bwc:minor")
+            .getBuildDir()
+            .toPath()
+            .resolve("bwc")
+            .resolve("checkout-" + priorMajorVersion + ".x");
```

Comment:
Why are we peeking at the source? Instead can we produce the files as an artifact? This would allow us to test against already released versions in the same way, just grab the zip of the spec/test files of that released version.

Reply:
The artifacts we produce only contain the OSS core api and tests. I need the x-pack core api/tests as well as all of the OSS/X-pack module/plugin API/tests. Back filling those I think is a reasonable approach as a long term solution, but I think sourcing like this is decent temporary approach.

Reply:
I tend to agree with @rjernst that peeking into other source folders is kind of an anti pattern. Instead sharing it declaratively as described here: https://docs.gradle.org/current/userguide/cross_project_publications.html is what we should aim for.

Reply:
Ideally I think the bwc project would just share the checkoutDir as an artifact. Can you put a TODO in there to look into this after https://github.com/elastic/elasticsearch/pull/62473 I will then pick that up at one. point. This PR does quite some rework on the bwc setup

Reply:
Thanks, exposing the checkoutDir as an artifact, I think will clean this up. I added a TODO in the code to come back to this.